### DISCUSSION TITLE
ST7789T3 ESP32

### DISCUSSION THREAD

**Discussion by bbuster2** [Bodmer/TFT_eSPI#3111] [New processor/display support or capability request]
I have bought a 1.83" ST7789T3 display but cannot get it working.

Display is bought here:
https://nl.aliexpress.com/item/1005005565821377.html?spm=a2g0o.order_list.order_list_main.11.205779d2rzxJEm&gatewayAdapt=glo2nld

Schematic:
![S94d8d0301cdc4382b0cfb085ce171f9f4](https://github.com/Bodmer/TFT_eSPI/assets/103888067/8b387966-408e-48fc-b1f0-2a0a54bad174)

In platformio, i use these biuld_flags:

```
build_flags = 
	-D USER_SETUP_LOADED=1
	-D ST7789_DRIVER=1
	-D SMOOTH_FONT=1
	-D TFT_MISO=-1
	-D TFT_MOSI=23
	-D TFT_SCLK=18
	-D TFT_CS=-1
	-D TFT_DC=2
	-D TFT_RST=4
	-D SPI_FREQUENCY=40000000
	-D SPI_READ_FREQUENCY=20000000
```

The display stays "black". I can control the backlight but i cannot get text or an image on the display.
I have the reset pin high (but also tried low), and the CS port connected to ground.

Is this display incompatible with the ESP32 or am i doing something wrong?

**Comment by Bodmer** [comment]
Try using the CS line on a GPIO pin. The reason is that the ST7789 is easy to lock up with garbage commands during power up.

Also run the Read_User_Setup sketch and check the output is correct, this shows whether the correct settings are being picked up.

**Comment by bbuster2** [comment]
Thank you for your reply. Adding the CS pin didn't make a difference (also connected to the ESP32).

Here is my output:
```
[code]
TFT_eSPI ver = 2.5.41
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = SPI
Display driver = 7789
Display width  = 240
Display height = 320

MOSI    = GPIO 23
SCK     = GPIO 18
TFT_CS   = GPIO 27
TFT_DC   = GPIO 2
TFT_RST  = GPIO 4

Font GLCD   loaded
Smooth font enabled

Display SPI frequency = 40.00
[/code]
```

**Comment by bbuster2** [comment]
Any idea?

**Comment by Bodmer** [comment]
The Read_User_Setup output looks good. Check the connections are sound and correct. Check the supply voltage to the display is correct.  Software options to try:

1. Use the alternative replacement driver, viz.  -D ST7789_2_DRIVER=1
2. Add option -D TFT_SPI_MODE SPI_MODE0=1
3. Reduce the SPI frequency to 27MHz

You also need to define the width and height to get the coordinate offsets correct, but that would not stop the display working.
-D TFT_WIDTH=240
-D TFT_HEIGHT=280

**Comment by bbuster2** [comment]
I tried but the same result.
I couldn't get -D TFT_SPI_MODE SPI_MODE0=1 to work
When i added that, i got the following error:

> *** [.pio/build/esp32dev/firmware.elf] Implicit dependency `/Users/{username}/.platformio/platforms/espressif32/builder/SPI_MODE0=1' not found, needed by target `.pio/build/esp32dev/firmware.elf'.

```
-D USER_SETUP_LOADED=1
	-D ST7789_2_DRIVER=1
	; -D TFT_SPI_MODE SPI_MODE0=1
	-D TFT_SPI_MODE=1
	-D SPI_MODE0=1

	-D DISPLAY_WIDTH=280
	-D DISPLAY_HEIGHT=240
	-D TFT_WIDTH=280
	-D TFT_HEIGHT=240
	-D SMOOTH_FONT=1
	-D TFT_MISO=-1
	-D TFT_MOSI=23
	-D TFT_SCLK=18
	-D TFT_CS=27
	-D TFT_DC=2
	-D TFT_RST=4
	-D TFT_WIDTH=240
	-D TFT_HEIGHT=280
	-D SPI_FREQUENCY=27000000
	-D SPI_READ_FREQUENCY=20000000
```

But i am not sure if that can work. The display is still black

**Comment by bbuster2** [comment]
My display looks really similar to this: https://www.buydisplay.com/1-83-inch-ips-tft-lcd-display-240x280-round-circle-screen-for-smart-watch

I see here the specifications show it is a **NV3030B** IC.
Can this work?

It is also 4 pin SPI

**Comment by Bodmer** [comment]
That display may look similar but that does not mean yours has that NV3030B controller. You have to assume the vendors information is correct.

I made a mistake above, the option should read:

-D TFT_SPI_MODE=SPI_MODE0

And delete the line:
-D SPI_MODE0=1

**Comment by bbuster2** [comment]
Thank you for your reply. Unfortunately i still cannot get anything on the display. It stays black

Maybe a really stupid question, but i don't see any chip on the display itself. It is just the screen with the 13 pin connector. Am i not missing something? (The driver chip?)

**Comment by bbuster2** [comment]
Any idea?

**Comment by Bodmer** [comment]
No further ideas on the problem.

I suggest you purchase a prebuilt board and get that running. That will help to identify the root cause (display/interface problem or library configuration issue.  If you have access to an oscilloscope or logic analyser that may help debug the problem too.

**Comment by nstepanets** [comment]
Most likely this display is on the NV3030B controller.

The controller is similar to ILI9341/GC9A01A but it seems that the initialization commands are different. I was able to get it working with ESP32 after using the code example from buydisplay.

**Reply by Gangwa-Labs** [reply]
can you post the code example from buydisplay? I downloaded the "8051 Interfacing and Demo Code" but can't seem to understand what I actually need to upload to my esp32

**Comment by nstepanets** [comment]
> Any idea?

@bbuster2 
If the question is still relevant, can you check the display with the changes from this pull request? 
https://github.com/Bodmer/TFT_eSPI/pull/3315

**Reply by Gangwa-Labs** [reply]
I have the exact same display and made the changes from this pull request and it did not work unfortunately and the screen is still full black

**Reply by nstepanets** [reply]
The screen is completely black, and the backlight is on?

If the display is on the ST7789 controller, then it should at least display something (with artifacts)

**Reply by Gangwa-Labs** [reply]
The screen is totally black without the backlight, shouldn't the display be on the new one from this pull request? NV3030B?

**Reply by nstepanets** [reply]
It looks like something is not set up correctly, as the backlight works regardless of the controller used.

**Comment by belpower01** [comment]
{
	LCD_RST = 1;
	Delayms(10);
	LCD_RST = 0;
	Delayms(10);
	LCD_RST = 1;
	Delayms(120);

NV3030B_MCU8_8080_CMD(0xfd);//private_access
NV3030B_MCU8_8080_DAT(0x06);
NV3030B_MCU8_8080_DAT(0x08);

NV3030B_MCU8_8080_CMD(0x61);//add
NV3030B_MCU8_8080_DAT(0x07);//
NV3030B_MCU8_8080_DAT(0x04);//
//bias
NV3030B_MCU8_8080_CMD(0x62);//bias setting
NV3030B_MCU8_8080_DAT(0x00);//01
NV3030B_MCU8_8080_DAT(0x44);//04 44
NV3030B_MCU8_8080_DAT(0x45);//44 65 40

NV3030B_MCU8_8080_CMD(0x63);//
NV3030B_MCU8_8080_DAT(0x41);//
NV3030B_MCU8_8080_DAT(0x07);//
NV3030B_MCU8_8080_DAT(0x12);//
NV3030B_MCU8_8080_DAT(0x12);//

NV3030B_MCU8_8080_CMD(0x64);//
NV3030B_MCU8_8080_DAT(0x37);//
//VSP
NV3030B_MCU8_8080_CMD(0x65);//Pump1=4.7MHz //PUMP1 VSP
NV3030B_MCU8_8080_DAT(0x09);//D6-5:pump1_clk[1:0] clamp 28 2b
NV3030B_MCU8_8080_DAT(0x10);//6.26
NV3030B_MCU8_8080_DAT(0x21);
//VSN
NV3030B_MCU8_8080_CMD(0x66); //pump=2 AVCL
NV3030B_MCU8_8080_DAT(0x09); //clamp 08 0b 09
NV3030B_MCU8_8080_DAT(0x10); //10
NV3030B_MCU8_8080_DAT(0x21);
//add source_neg_time
NV3030B_MCU8_8080_CMD(0x67);//pump_sel
NV3030B_MCU8_8080_DAT(0x20);//21 20
NV3030B_MCU8_8080_DAT(0x40);
//gamma vap/van
NV3030B_MCU8_8080_CMD(0x68);//gamma vap/van
NV3030B_MCU8_8080_DAT(0x90);//
NV3030B_MCU8_8080_DAT(0x4c);//
NV3030B_MCU8_8080_DAT(0x7c);//
NV3030B_MCU8_8080_DAT(0x66);//66  46  VCOM

NV3030B_MCU8_8080_CMD(0xb1);//frame rate
NV3030B_MCU8_8080_DAT(0x0F);//0x0f fr_h[5:0] 0F
NV3030B_MCU8_8080_DAT(0x02);//0x02 fr_v[4:0] 02
NV3030B_MCU8_8080_DAT(0x01);//0x04 fr_div[2:0] 03

NV3030B_MCU8_8080_CMD(0xB4);
NV3030B_MCU8_8080_DAT(0x01); //01:1dot 00:column
////porch
NV3030B_MCU8_8080_CMD(0xB5);
NV3030B_MCU8_8080_DAT(0x02);//0x02 vfp[6:0]
NV3030B_MCU8_8080_DAT(0x02);//0x02 vbp[6:0]
NV3030B_MCU8_8080_DAT(0x0a);//0x0A hfp[6:0]
NV3030B_MCU8_8080_DAT(0x14);//0x14 hbp[6:0]

NV3030B_MCU8_8080_CMD(0xB6);
NV3030B_MCU8_8080_DAT(0x04);//
NV3030B_MCU8_8080_DAT(0x01);//
NV3030B_MCU8_8080_DAT(0x9f);//
NV3030B_MCU8_8080_DAT(0x00);//
NV3030B_MCU8_8080_DAT(0x02);//
////gamme sel
NV3030B_MCU8_8080_CMD(0xdf);//
NV3030B_MCU8_8080_DAT(0x11);//gofc_gamma_en_sel=1
////gamma_test1 A1#_wangly
//3030b_gamma_new_
//GAMMA---------------------------------/////////////

//GAMMA---------------------------------/////////////
NV3030B_MCU8_8080_CMD(0xE2);	
NV3030B_MCU8_8080_DAT(0x13);//vrp0[5:0]	V0
NV3030B_MCU8_8080_DAT(0x00);//vrp1[5:0]	V1  
NV3030B_MCU8_8080_DAT(0x00);//vrp2[5:0]	V2 
NV3030B_MCU8_8080_DAT(0x30);//vrp3[5:0]	V61 
NV3030B_MCU8_8080_DAT(0x33);//vrp4[5:0]	V62
NV3030B_MCU8_8080_DAT(0x3f);//vrp5[5:0]	V63

NV3030B_MCU8_8080_CMD(0xE5);	
NV3030B_MCU8_8080_DAT(0x3f);//vrn0[5:0]	V63
NV3030B_MCU8_8080_DAT(0x33);//vrn1[5:0]	V62	
NV3030B_MCU8_8080_DAT(0x30);//vrn2[5:0]	V61 
NV3030B_MCU8_8080_DAT(0x00);//vrn3[5:0]	V2 
NV3030B_MCU8_8080_DAT(0x00);//vrn4[5:0]	V1	
NV3030B_MCU8_8080_DAT(0x13);//vrn5[5:0]  V0

NV3030B_MCU8_8080_CMD(0xE1);	
NV3030B_MCU8_8080_DAT(0x00);//prp0[6:0]	V15
NV3030B_MCU8_8080_DAT(0x57);//prp1[6:0]	V51 

NV3030B_MCU8_8080_CMD(0xE4);	
NV3030B_MCU8_8080_DAT(0x58);//prn0[6:0]	V51 
NV3030B_MCU8_8080_DAT(0x00);//prn1[6:0]  V15

NV3030B_MCU8_8080_CMD(0xE0);
NV3030B_MCU8_8080_DAT(0x01);//pkp0[4:0]	V3 
NV3030B_MCU8_8080_DAT(0x03);//pkp1[4:0]	V7  
NV3030B_MCU8_8080_DAT(0x0d);//pkp2[4:0]	V21 
NV3030B_MCU8_8080_DAT(0x0e);//pkp3[4:0]	V29 //
NV3030B_MCU8_8080_DAT(0x0e);//pkp4[4:0]	V37 
NV3030B_MCU8_8080_DAT(0x0c);//pkp5[4:0]	V45 
NV3030B_MCU8_8080_DAT(0x15);//pkp6[4:0]	V56 
NV3030B_MCU8_8080_DAT(0x19);//pkp7[4:0]	V60 

NV3030B_MCU8_8080_CMD(0xE3);	
NV3030B_MCU8_8080_DAT(0x1a);//pkn0[4:0]	V60 
NV3030B_MCU8_8080_DAT(0x16);//pkn1[4:0]	V56 
NV3030B_MCU8_8080_DAT(0x0c);//pkn2[4:0]	V45 
NV3030B_MCU8_8080_DAT(0x0f);//pkn3[4:0]	V37 
NV3030B_MCU8_8080_DAT(0x0e);//pkn4[4:0]	V29 //
NV3030B_MCU8_8080_DAT(0x0d);//pkn5[4:0]	V21 
NV3030B_MCU8_8080_DAT(0x02);//pkn6[4:0]	V7  
NV3030B_MCU8_8080_DAT(0x01);//pkn7[4:0]	V3 
//GAMMA---------------------------------/////////////


//source
NV3030B_MCU8_8080_CMD(0xE6);
NV3030B_MCU8_8080_DAT(0x00);
NV3030B_MCU8_8080_DAT(0xff);//SC_EN_START[7:0] f0

NV3030B_MCU8_8080_CMD(0xE7);
NV3030B_MCU8_8080_DAT(0x01);//CS_START[3:0] 01
NV3030B_MCU8_8080_DAT(0x04);//scdt_inv_sel cs_vp_en
NV3030B_MCU8_8080_DAT(0x03);//CS1_WIDTH[7:0] 12
NV3030B_MCU8_8080_DAT(0x03);//CS2_WIDTH[7:0] 12
NV3030B_MCU8_8080_DAT(0x00);//PREC_START[7:0] 06
NV3030B_MCU8_8080_DAT(0x12);//PREC_WIDTH[7:0] 12

NV3030B_MCU8_8080_CMD(0xE8); //source
NV3030B_MCU8_8080_DAT(0x00); //VCMP_OUT_EN 81-
NV3030B_MCU8_8080_DAT(0x70); //chopper_sel[6:4]
NV3030B_MCU8_8080_DAT(0x00); //gchopper_sel[6:4] 60
////gate
NV3030B_MCU8_8080_CMD(0xEc);
NV3030B_MCU8_8080_DAT(0x52);//52

NV3030B_MCU8_8080_CMD(0xF1);
NV3030B_MCU8_8080_DAT(0x01);//te_pol tem_extend 00 01 03
NV3030B_MCU8_8080_DAT(0x01);
NV3030B_MCU8_8080_DAT(0x02);

NV3030B_MCU8_8080_CMD(0xF6);//
NV3030B_MCU8_8080_DAT(0x09);//
NV3030B_MCU8_8080_DAT(0x10);//
NV3030B_MCU8_8080_DAT(0x00);//
NV3030B_MCU8_8080_DAT(0x00);//40 3ПЯ2НЁµА

NV3030B_MCU8_8080_CMD(0xfd);
NV3030B_MCU8_8080_DAT(0xfa);
NV3030B_MCU8_8080_DAT(0xfc);

NV3030B_MCU8_8080_CMD(0x3a);
NV3030B_MCU8_8080_DAT(0x05);//SH 0x66

NV3030B_MCU8_8080_CMD(0x35);
NV3030B_MCU8_8080_DAT(0x00);

NV3030B_MCU8_8080_CMD(0x36);//bgr_[3]
NV3030B_MCU8_8080_DAT(0x00);//c0

NV3030B_MCU8_8080_CMD(0x21); 

NV3030B_MCU8_8080_CMD(0x11); // exit sleep
Delayms(200);
NV3030B_MCU8_8080_CMD(0x29); // display on
Delayms(10);
}


/*====================================================*/
void address(void)
{             
	NV3030B_MCU8_8080_CMD(0x2A);  //Column Address Set
	NV3030B_MCU8_8080_DAT(0x00);  
	NV3030B_MCU8_8080_DAT(0x00);  
	NV3030B_MCU8_8080_DAT(0x00);
	NV3030B_MCU8_8080_DAT(0xEF);  /*  240  */

	NV3030B_MCU8_8080_CMD(0x2B);   //Row Address Set   
	NV3030B_MCU8_8080_DAT(0x00);
	NV3030B_MCU8_8080_DAT(0x14);    /*  20  */
	NV3030B_MCU8_8080_DAT(0x01);
	NV3030B_MCU8_8080_DAT(0x2B);   /*  299  */
				   
	NV3030B_MCU8_8080_CMD(0x2C);
 }

/**********SLEEP IN***************/
void EnterSleep(void)
{
	NV3030B_MCU8_8080_CMD(0x28);
	NV3030B_MCU8_8080_CMD(0x10);
	Delayms(120);
}

/**********SLEEP OUT***************/
void ExitSleep(void)
{

	NV3030B_MCU8_8080_CMD(0x11);
	Delayms(120);
	NV3030B_MCU8_8080_CMD(0x29);
        Delayms(10);
}

**Reply by nstepanets** [reply]
```
NV3030B_MCU8_8080_CMD(0x3a);
NV3030B_MCU8_8080_DAT(0x05);//SH 0x66
```
According to the datasheet, the correct value is 0x55

```
NV3030B_MCU8_8080_CMD(0x35);
NV3030B_MCU8_8080_DAT(0x00);
```
Tearing effect line on - I'm not sure that modes that are not used on other displays should be enabled.

```
NV3030B_MCU8_8080_CMD(0x36);//bgr_[3]
NV3030B_MCU8_8080_DAT(0x00);//c0
```
This is the default value, there is no point in setting it again.

**Comment by belpower01** [comment]
void LCD_GPIO_Init(void)
{
	GPIO_InitTypeDef GPIO_InitStructure;
	RCC_AHBPeriphClockCmd(RCC_AHBPeriph_GPIOA, ENABLE);
 
	GPIO_InitStructure.GPIO_Pin = GPIO_Pin_0|GPIO_Pin_1|GPIO_Pin_2|GPIO_Pin_3|GPIO_Pin_4|GPIO_Pin_5;
	GPIO_InitStructure.GPIO_Speed = GPIO_Speed_50MHz;
	GPIO_InitStructure.GPIO_Mode = GPIO_Mode_OUT;
	GPIO_InitStructure.GPIO_OType = GPIO_OType_PP;
	GPIO_Init(GPIOA, &GPIO_InitStructure);
	GPIO_SetBits(GPIOA,GPIO_Pin_0|GPIO_Pin_1|GPIO_Pin_2|GPIO_Pin_3|GPIO_Pin_4|GPIO_Pin_5);
}


void LCD_Writ_Bus(u8 dat) 
{	
	u8 i;
	LCD_CS_Clr();
	for(i=0;i<8;i++)
	{			  
		LCD_SCLK_Clr();
		if(dat&0x80)
		{
		   LCD_MOSI_Set();
		}
		else
		{
		   LCD_MOSI_Clr();
		}
		LCD_SCLK_Set();
		dat<<=1;
	}	
  LCD_CS_Set();	
}



void LCD_WR_DATA8(u8 dat)
{
	LCD_Writ_Bus(dat);
}


void LCD_WR_DATA(u16 dat)
{
	LCD_Writ_Bus(dat>>8);
	LCD_Writ_Bus(dat);
}


void LCD_WR_REG(u8 dat)
{
	LCD_DC_Clr();
	LCD_Writ_Bus(dat);
	LCD_DC_Set();
}


void LCD_Address_Set(u16 x1,u16 y1,u16 x2,u16 y2)
{
	if(USE_HORIZONTAL==0)
	{
		LCD_WR_REG(0x2a);
		LCD_WR_DATA(x1);
		LCD_WR_DATA(x2);
		LCD_WR_REG(0x2b);
		LCD_WR_DATA(y1+20);
		LCD_WR_DATA(y2+20);
		LCD_WR_REG(0x2c);
	}
	else if(USE_HORIZONTAL==1)
	{
		LCD_WR_REG(0x2a);
		LCD_WR_DATA(x1);
		LCD_WR_DATA(x2);
		LCD_WR_REG(0x2b);
		LCD_WR_DATA(y1+20);
		LCD_WR_DATA(y2+20);
		LCD_WR_REG(0x2c);
	}
	else if(USE_HORIZONTAL==2)
	{
		LCD_WR_REG(0x2a);
		LCD_WR_DATA(x1+20);
		LCD_WR_DATA(x2+20);
		LCD_WR_REG(0x2b);
		LCD_WR_DATA(y1);
		LCD_WR_DATA(y2);
		LCD_WR_REG(0x2c);
	}
	else
	{
		LCD_WR_REG(0x2a);
		LCD_WR_DATA(x1+20);
		LCD_WR_DATA(x2+20);
		LCD_WR_REG(0x2b);
		LCD_WR_DATA(y1);
		LCD_WR_DATA(y2);
		LCD_WR_REG(0x2c);
	}
}
void LCD_Init(void)
{
	LCD_GPIO_Init();
	
	LCD_RES_Set();
	delay_ms(10);
	LCD_RES_Clr();
	delay_ms(10);
	LCD_RES_Set();
	delay_ms(120);
	
	LCD_BLK_Set();
	
	LCD_WR_REG(0xfd);
	LCD_WR_DATA8(0x06);
	LCD_WR_DATA8(0x08);

	LCD_WR_REG(0x61);
	LCD_WR_DATA8(0x07);
	LCD_WR_DATA8(0x04);

	LCD_WR_REG(0x62);
	LCD_WR_DATA8(0x00);
	LCD_WR_DATA8(0x44);
	LCD_WR_DATA8(0x45);

	LCD_WR_REG(0x63);
	LCD_WR_DATA8(0x41);
	LCD_WR_DATA8(0x07);
	LCD_WR_DATA8(0x12);
	LCD_WR_DATA8(0x12);

	LCD_WR_REG(0x64);
	LCD_WR_DATA8(0x37);
	//VSP
	LCD_WR_REG(0x65);
	LCD_WR_DATA8(0x09);
	LCD_WR_DATA8(0x10);
	LCD_WR_DATA8(0x21);
	//VSN
	LCD_WR_REG(0x66); 
	LCD_WR_DATA8(0x09); 
	LCD_WR_DATA8(0x10); 
	LCD_WR_DATA8(0x21);
	//add source_neg_time
	LCD_WR_REG(0x67);//pump_sel
	LCD_WR_DATA8(0x20);//21 20
	LCD_WR_DATA8(0x40);

	//gamma vap/van
	LCD_WR_REG(0x68);//gamma vap/van
	LCD_WR_DATA8(0x90);//
	LCD_WR_DATA8(0x4c);//
	LCD_WR_DATA8(0x7C);//VCOM  
	LCD_WR_DATA8(0x66);//

	LCD_WR_REG(0xb1);//frame rate
	LCD_WR_DATA8(0x0F);//0x0f fr_h[5:0] 0F
	LCD_WR_DATA8(0x02);//0x02 fr_v[4:0] 02
	LCD_WR_DATA8(0x01);//0x04 fr_div[2:0] 04

	LCD_WR_REG(0xB4);
	LCD_WR_DATA8(0x01); //01:1dot 00:column
	////porch
	LCD_WR_REG(0xB5);
	LCD_WR_DATA8(0x02);//0x02 vfp[6:0]
	LCD_WR_DATA8(0x02);//0x02 vbp[6:0]
	LCD_WR_DATA8(0x0a);//0x0A hfp[6:0]
	LCD_WR_DATA8(0x14);//0x14 hbp[6:0]

	LCD_WR_REG(0xB6);
	LCD_WR_DATA8(0x04);//
	LCD_WR_DATA8(0x01);//
	LCD_WR_DATA8(0x9f);//
	LCD_WR_DATA8(0x00);//
	LCD_WR_DATA8(0x02);//
	////gamme sel
	LCD_WR_REG(0xdf);//
	LCD_WR_DATA8(0x11);//gofc_gamma_en_sel=1
	////gamma_test1 A1#_wangly
	//3030b_gamma_new_
	//GAMMA---------------------------------/////////////

	//GAMMA---------------------------------/////////////
	LCD_WR_REG(0xE2);	
	LCD_WR_DATA8(0x13);//vrp0[5:0]	V0 13
	LCD_WR_DATA8(0x00);//vrp1[5:0]	V1 
	LCD_WR_DATA8(0x00);//vrp2[5:0]	V2 
	LCD_WR_DATA8(0x30);//vrp3[5:0]	V61 
	LCD_WR_DATA8(0x33);//vrp4[5:0]	V62 
	LCD_WR_DATA8(0x3f);//vrp5[5:0]	V63

	LCD_WR_REG(0xE5);	
	LCD_WR_DATA8(0x3f);//vrn0[5:0]	V63
	LCD_WR_DATA8(0x33);//vrn1[5:0]	V62	
	LCD_WR_DATA8(0x30);//vrn2[5:0]	V61 
	LCD_WR_DATA8(0x00);//vrn3[5:0]	V2 
	LCD_WR_DATA8(0x00);//vrn4[5:0]	V1 
	LCD_WR_DATA8(0x13);//vrn5[5:0]  V0 13

	LCD_WR_REG(0xE1);	
	LCD_WR_DATA8(0x00);//prp0[6:0]	V15
	LCD_WR_DATA8(0x57);//prp1[6:0]	V51 

	LCD_WR_REG(0xE4);	
	LCD_WR_DATA8(0x58);//prn0[6:0]	V51 
	LCD_WR_DATA8(0x00);//prn1[6:0]  V15

	LCD_WR_REG(0xE0);
	LCD_WR_DATA8(0x01);//pkp0[4:0]	V3 
	LCD_WR_DATA8(0x03);//pkp1[4:0]	V7  
	LCD_WR_DATA8(0x0d);//pkp2[4:0]	V21
	LCD_WR_DATA8(0x0e);//pkp3[4:0]	V29 
	LCD_WR_DATA8(0x0e);//pkp4[4:0]	V37 
	LCD_WR_DATA8(0x0c);//pkp5[4:0]	V45 
	LCD_WR_DATA8(0x15);//pkp6[4:0]	V56 
	LCD_WR_DATA8(0x19);//pkp7[4:0]	V60 

	LCD_WR_REG(0xE3);	
	LCD_WR_DATA8(0x1a);//pkn0[4:0]	V60 
	LCD_WR_DATA8(0x16);//pkn1[4:0]	V56 
	LCD_WR_DATA8(0x0C);//pkn2[4:0]	V45 
	LCD_WR_DATA8(0x0f);//pkn3[4:0]	V37 
	LCD_WR_DATA8(0x0e);//pkn4[4:0]	V29 
	LCD_WR_DATA8(0x0d);//pkn5[4:0]	V21 
	LCD_WR_DATA8(0x02);//pkn6[4:0]	V7  
	LCD_WR_DATA8(0x01);//pkn7[4:0]	V3 
	//GAMMA---------------------------------/////////////

	//source
	LCD_WR_REG(0xE6);
	LCD_WR_DATA8(0x00);
	LCD_WR_DATA8(0xff);//SC_EN_START[7:0] f0

	LCD_WR_REG(0xE7);
	LCD_WR_DATA8(0x01);//CS_START[3:0] 01
	LCD_WR_DATA8(0x04);//scdt_inv_sel cs_vp_en
	LCD_WR_DATA8(0x03);//CS1_WIDTH[7:0] 12
	LCD_WR_DATA8(0x03);//CS2_WIDTH[7:0] 12
	LCD_WR_DATA8(0x00);//PREC_START[7:0] 06
	LCD_WR_DATA8(0x12);//PREC_WIDTH[7:0] 12

	LCD_WR_REG(0xE8); //source
	LCD_WR_DATA8(0x00); //VCMP_OUT_EN 81-
	LCD_WR_DATA8(0x70); //chopper_sel[6:4]
	LCD_WR_DATA8(0x00); //gchopper_sel[6:4] 60
	////gate
	LCD_WR_REG(0xEc);
	LCD_WR_DATA8(0x52);//52

	LCD_WR_REG(0xF1);
	LCD_WR_DATA8(0x01);//te_pol tem_extend 00 01 03
	LCD_WR_DATA8(0x01);
	LCD_WR_DATA8(0x02);


	LCD_WR_REG(0xF6);
	LCD_WR_DATA8(0x09);
	LCD_WR_DATA8(0x10);
	LCD_WR_DATA8(0x00);//
	LCD_WR_DATA8(0x00);//40 3ПЯ2НЁµА

	LCD_WR_REG(0xfd);
	LCD_WR_DATA8(0xfa);
	LCD_WR_DATA8(0xfc);

	LCD_WR_REG(0x3a);
	LCD_WR_DATA8(0x05);//

	LCD_WR_REG(0x35);
	LCD_WR_DATA8(0x00);

	LCD_WR_REG(0x36);
	if(USE_HORIZONTAL==0)LCD_WR_DATA8(0x08);
	else if(USE_HORIZONTAL==1)LCD_WR_DATA8(0xC8);
	else if(USE_HORIZONTAL==2)LCD_WR_DATA8(0x78);
	else LCD_WR_DATA8(0xA8);


	LCD_WR_REG(0x21); 

	LCD_WR_REG(0x11); // exit sleep
	delay_ms(200);
	LCD_WR_REG(0x29); // display on
	delay_ms(10); 
}

**Comment by kocser2** [comment]
I have the same problem with the same screen from the same vendor. I have contacted the vendor but have not yet received a reply.
Has anyone managed to bring this screen to life?

**Reply by nstepanets** [reply]
Try changing the SPI mode in the display configuration:
#define TFT_SPI_MODE SPI_MODE0

If this does not help, then perhaps the display uses the NV3030B controller

**Reply by max22-** [reply]
Hi,
i don't have this specific display, but a [Waveshare RP2350-Touch-LCD-2](https://www.waveshare.com/wiki/RP2350-Touch-LCD-2). It is also based on a ST7789T3 driver. I've adapted the code from the example provided by waveshare, and integrated it in TFT_eSPI (i've just made a pull request).
The code is here : [https://github.com/max22-/TFT_eSPI/tree/ST7789T3](https://github.com/max22-/TFT_eSPI/tree/ST7789T3)
(it is in the ST7789T3 branch).
Maybe it could be useful for your screen as well.
